### PR TITLE
Osv: Somewhat improve the mapping of the severity

### PR DESCRIPTION
**Pull request by fviernau**
ORT's data model for the severity of a vulnerability is quite generic.
The wide flexibiltiy it provides makes it unclear how to best map the
obtained values from the provider to that data model.

The use of `VulnerabilityReference.severity` in [1] indicates that a
float score value is a valid option. So, change the mapping to float for
now as that is easier to understand - even though this is a lossy
transformation.

In the longer run, ORT's vulnerability data model should be improved to
support vectors in a less generic stricter typed way.

[1] https://github.com/oss-review-toolkit/ort/blob/470e2f323b516389bd96a63a81c67b74891a36f8/model/src/main/kotlin/VulnerabilityReference.kt#L58-L63

This is part of #3599.

### CODE REVIEW

File: advisor/src/funTest/assets/retrieve-package-findings-expected-result.json [hunk 1]
```diff
@@ -13,41 +13,41 @@
...
         "url" : "https://github.com/gulpjs/glob-parent",
-        "scoring_system" : "CVSS_V3",
-        "severity" : "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
+        "scoring_system" : "CVSS:3.1",
+        "severity" : "7.5"
       } ]
     } ]
   } ]
-}
+}
```
Comment (tsteenbe) [inline advisor/src/funTest/assets/retrieve-package-findings-expected-result.json:53]: Missing new line

File: advisor/src/main/kotlin/advisors/Osv.kt [hunk 2]
```diff
@@ -168,19 +171,30 @@ private fun createRequest(pkg: Package): VulnerabilitiesForPackageRequest? {
 }
 
 private fun Vulnerability.toOrtVulnerability(): org.ossreviewtoolkit.model.Vulnerability {
-    val (scoringSystem, severity) = severity.firstOrNull()?.let {
-        it.type.name to it.score
-    } ?: null to null
+    val (scoringSystem, score) = severity.firstOrNull()?.let {
+        val cvss = Cvss.fromVector(it.score) ?: run {
+            log.debug { "Could not parse CVSS vector from: '${it.score}'." }
+            return@let null to it.score
```
Comment (sschuberth) [inline advisor/src/main/kotlin/advisors/Osv.kt:177]: This avoids the IMO not very readable `return@` construct:

        Cvss.fromVector(it.score)?.let { cvss ->
            val scoringSystem = when {
                // Work around for https://github.com/stevespringett/cvss-calculator/issues/56.
                it.score.startsWith("CVSS:") -> it.score.substringBefore("/")
                cvss is CvssV2 -> "CVSS:2.0"
                else -> cvss.vector.substringBefore("/", "CVSS")
            }

            scoringSystem to cvss.calculateScore().baseScore.toFloat().toString()
        } ?: run {
            log.debug { "Could not parse CVSS vector from: '${it.score}'." }
            null to it.score
        }

File: advisor/src/main/kotlin/advisors/Osv.kt [hunk 3]
```diff
@@ -168,19 +171,30 @@ private fun createRequest(pkg: Package): VulnerabilitiesForPackageRequest? {
 }
 
 private fun Vulnerability.toOrtVulnerability(): org.ossreviewtoolkit.model.Vulnerability {
-    val (scoringSystem, severity) = severity.firstOrNull()?.let {
-        it.type.name to it.score
-    } ?: null to null
+    val (scoringSystem, score) = severity.firstOrNull()?.let {
+        val cvss = Cvss.fromVector(it.score) ?: run {
+            log.debug { "Could not parse CVSS vector from: '${it.score}'." }
```
Comment (sschuberth) [inline advisor/src/main/kotlin/advisors/Osv.kt:176]: This reads as if you're trying to *create* a vector from `it.score`. Just say "Could not parse CVSS vector '${it.score}'."

File: advisor/src/main/kotlin/advisors/Osv.kt [hunk 4]
```diff
@@ -168,19 +171,30 @@ private fun createRequest(pkg: Package): VulnerabilitiesForPackageRequest? {
...
+
+        val scoringSystem = when {
+            // Work around for https://github.com/stevespringett/cvss-calculator/issues/56.
+            it.score.startsWith("CVSS:") -> it.score.substringBefore("/")
+            cvss is CvssV2 -> "CVSS:2.0"
+            else -> cvss.vector.substringBefore("/", "CVSS")
+        }
+
+        scoringSystem to cvss.calculateScore().baseScore.toFloat().toString()
+    } ?: (null to null)
```
Comment (sschuberth) [inline advisor/src/main/kotlin/advisors/Osv.kt:188]: Nit: This inspection hint could be fixed in a separate commit to reduce the diff here.

File: advisor/src/main/kotlin/advisors/Osv.kt [hunk 5]
```diff
@@ -168,12 +171,23 @@ private fun createRequest(pkg: Package): VulnerabilitiesForPackageRequest? {
...
+    val (scoringSystem, severity) = this.severity.firstOrNull()?.let {
+        Cvss.fromVector(it.score)?.let { cvss ->
+            val scoringSystem = when {
+                // Work around for https://github.com/stevespringett/cvss-calculator/issues/56.
+                it.score.startsWith("CVSS:") -> it.score.substringBefore("/")
+                cvss is CvssV2 -> "CVSS:2.0"
+                else -> cvss.vector.substringBefore("/", "CVSS")
+            }
+
+            scoringSystem to cvss.calculateScore().baseScore.toFloat().toString()
```
Comment (sschuberth) [inline advisor/src/main/kotlin/advisors/Osv.kt:183]: Nit: Maybe just do "${cvss.calculateScore().baseScore}"?

**Review by sschuberth — APPROVED** [review]
Besides the optional comment fine with me.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**